src/platforms/stm32f4-discovery.c

```diff
@@ -1059,62 +990,175 @@ void dma1_stream3_isr(void) {
       adc_value >>= 4; /* 12 bit value is left justified */
 #endif
       config.sensors[i].raw_value = adc_value;
+    } else if (config.sensors[i].source == SENSOR_FREQ) {
+      uint8_t pin = config.sensors[i].pin;
+      if (pin > 3) {
+        continue;
+      }
+
+      /* TODO more intelligence, for now last sample time diff */
+      ssize_t before = freq_input_position(pin) - 2;
+      before = before < 0 ? FREQ_INPUT_SAMPLES + before : before;
+
+      ssize_t after = freq_input_position(pin) - 1;
```

Comment:
This is racy, we should pick before, and then pick after to be before+1, in case multiple samples happen during those two points.